pycrostates/cluster/aahc.py

```diff
@@ -0,0 +1,408 @@
+        assert self.GEV_ is not None
+
+    @copy_doc(_BaseCluster.fit)
+    @fill_doc
+    def fit(
+        self,
+        inst: Union[BaseRaw, BaseEpochs],
+        picks: Picks = "eeg",
+        tmin: Optional[Union[int, float]] = None,
+        tmax: Optional[Union[int, float]] = None,
+        reject_by_annotation: bool = True,
+        n_jobs: int = 1,
+        *,
+        verbose: Optional[str] = None,
+    ) -> None:
+        """
+        %(verbose)s
+        """
+        _set_verbose(verbose)  # TODO: decorator nesting is failing
+        data = super().fit(
+            inst, picks, tmin, tmax, reject_by_annotation, n_jobs
+        )
+
+        gev, maps, segmentation = AAHCluster._aahc(
+            data,
+            self._n_clusters,
+            self._ignore_polarity,
+            self._normalize_input,
+            self._tol,
+        )
+
+        if gev is not None:
+            logger.info("AAHC converged with GEV = %.2f%% ", gev * 100)
+
+        self._GEV_ = gev
+        self._cluster_centers_ = maps
+        self._labels_ = segmentation
+        self._fitted = True
+
+    @copy_doc(_BaseCluster.save)
+    def save(self, fname: Union[str, Path]):
+        super().save(fname)
+        # TODO: to be replaced by a general writer than infers the writer from
+        # the file extension.
+        from ..io.fiff import _write_cluster  # pylint: disable=C0415
+
+        _write_cluster(
+            fname,
+            self._cluster_centers_,
+            self._info,
+            "AAHCluster",
+            self._cluster_names,
+            self._fitted_data,
+            self._labels_,
+            ignore_polarity=self._ignore_polarity,
+            normalize_input=self._normalize_input,
+            tol=self._tol,
+            GEV_=self._GEV_,
+        )
+
+    # --------------------------------------------------------------------
+    @staticmethod
+    def _aahc(
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+        tol: Union[int, float],
+    ) -> Tuple[float, NDArray[float], NDArray[int]]:
+        """Run the AAHC algorithm."""
+        gfp_sum_sq = np.sum(data**2)
+        maps, segmentation = AAHCluster._compute_maps(
+            data, n_clusters, ignore_polarity, normalize_input, tol
+        )
+        map_corr = _corr_vectors(data, maps[segmentation].T)
+        gev = np.sum((data * map_corr) ** 2) / gfp_sum_sq
+        return gev, maps, segmentation
+
+    @staticmethod
+    def _compute_maps(
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+        tol: Union[int, float],
+    ) -> Tuple[NDArray[float], NDArray[int]]:
+        """Compute microstates maps."""
+        n_chan, n_frame = data.shape
+
+        cluster = data.copy()
+        cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
+
+        if normalize_input:
+            data = cluster.copy()
```

Comment:
Could you add a test with `normalize_input=True`?

Reply:
Yes, I will add a test case.

Reply:
Resolved in https://github.com/vferat/pycrostates/pull/92/commits/a3f7afc062b2f43cf61a9bc2b43f7af29a470524

Reply:
Resolved in commit https://github.com/vferat/pycrostates/pull/92/commits/29fa0e53ffd2a1aefc633a32e47bec430afcb275